types/src/checksummed_hex.rs

```diff
@@ -0,0 +1,584 @@
+    let mut maybe_hash_bits =
+        (input_bytes.len() <= SMALL_BYTES_COUNT).then(|| bytes_to_bits_cycle(blake2b_hash(input)));
+    nibbles.map(move |mut nibble| {
+        if let Some(hash_bits) = maybe_hash_bits.as_mut() {
+            if nibble >= 10 && hash_bits.next().unwrap_or(true) {
+                nibble += 6;
+            }
+        }
+        HEX_CHARS[nibble as usize]
+    })
+}
+
+/// Returns true if all chars in a string are uppercase or lowercase.
+/// Returns false if the string is mixed case or if there are no alphabetic chars.
+fn string_is_same_case<T: AsRef<[u8]> + ?Sized>(s: &T) -> bool {
+    let mut chars = s.as_ref().iter().filter(|c| c.is_ascii_alphabetic());
+
+    match chars.next() {
+        Some(first) => {
+            let is_upper = first.is_ascii_uppercase();
+            chars.all(|c| c.is_ascii_uppercase() == is_upper)
+        }
+        None => {
+            // String has no actual characters.
+            true
+        }
+    }
+}
+
+/// Decodes a mixed-case hexadecimal string, verifying that it conforms to the checksum scheme
+/// similar to scheme in [EIP-55][1].
+///
+/// Key differences:
+///   - Works on any length of data, not just 20-byte addresses
+///   - Uses Blake2b hashes rather than Keccak
+///   - Hash nibbles based on input rather than hex encoding of input
+///   - Hash nibbles are XORed with input nibbles
+///
+/// For backward compatibility: if the hex string is all uppercase or all lowercase, the check is
+/// skipped.
+///
+/// [1]: https://eips.ethereum.org/EIPS/eip-55
+pub fn decode(input: &(impl AsRef<[u8]> + ?Sized)) -> Result<Vec<u8>, base16::DecodeError> {
+    let bytes = base16::decode(input)?;
+    // If the string was all uppercase or all lower case, don't verify the checksum.
+    // This is to support legacy clients.
+    // Otherwise perform the check as below.
+    if !string_is_same_case(input.as_ref()) {
+        let input_string_bytes = input.as_ref();
+
+        encode_iter(&bytes)
+            .zip(input_string_bytes.iter())
+            .enumerate()
+            .try_for_each(|(index, (expected_case_hex_char, &input_hex_char))| {
+                if expected_case_hex_char as u8 == input_hex_char {
+                    Ok(())
+                } else {
+                    Err(base16::DecodeError::InvalidByte {
+                        index,
+                        byte: expected_case_hex_char as u8,
+                    })
+                }
+            })?;
+    }
+    Ok(bytes)
+}
+
+/// Provides checksummed hex-encoded (de)serialization for `serde` when used with human-readable
+/// (de/en)coders, following an [EIP-55][1]-like scheme.  For non-human-readable (de/en)coders, the
+/// value is (de)serialized as a byte array.
+///
+/// Note that the trait is automatically implemented for types that implement [`AsRef`]`<[u8]>` and
+/// [`TryFrom`]`<&[u8]>`.
+///
+/// [1]: https://eips.ethereum.org/EIPS/eip-55
+pub trait CheckSummedHex<T> {
```

Comment:
```suggestion
pub trait ChecksummedHex<T> {
```

My bad for messing up the capitalization here :(